pkg/proxy/iptables/proxier.go

```diff
@@ -1472,25 +1499,64 @@ func (proxier *Proxier) syncProxyRules() {
 				}
 			}
 
-			// Setup probability filter rules only over local endpoints
-			for i, endpointChain := range localEndpointChains {
+			for i, endpointChain := range localReadyEndpointChains {
 				// Balancing rules in the per-service chain.
 				args = append(args[:0],
 					"-A", string(svcXlbChain),
 					"-m", "comment", "--comment",
 					fmt.Sprintf(`"Balancing rule %d for %s"`, i, svcNameString),
 				)
-				if i < (numLocalEndpoints - 1) {
+				if i < (numLocalReadyEndpoints - 1) {
 					// Each rule is a probabilistic match.
 					args = append(args,
 						"-m", "statistic",
 						"--mode", "random",
-						"--probability", proxier.probability(numLocalEndpoints-i))
+						"--probability", proxier.probability(numLocalReadyEndpoints-i))
 				}
 				// The final (or only if n == 1) rule is a guaranteed match.
 				args = append(args, "-j", string(endpointChain))
 				writeLine(proxier.natRules, args...)
 			}
+		} else if len(localReadyTerminatingEndpointChains) > 0 {
+			if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
```

Comment:
the `localReadyEndpointChains` and `localReadyTerminatingEndpointChains` cases are identical except for which array they're looking at, and they're very very similar to the `readyEndpointChains` case above... some of this code should be abstracted out into a function

although also, do you really need to keep track of both `localReadyEndpointChains` and `localReadyTerminatingEndpointChains`? Couldn't you just end the "categorize" loop by doing

    localEndpointChains := localReadyEndpointChains
    if len(localEndpointChains) == 0 {
            localEndpointChains = localReadyTerminatingEndpointChains
    }

?

Reply:
Yeah, great suggestion. Will update